### ISSUE TITLE
How to obtain the training epoch in customized loss function?

### ISSUE THREAD

**Issue by Cogito2012** [open-mmlab/mmaction2#475]
For some kinds of loss functions, the weighting factor could be an **annealing coefficient depending on the training epoch**. Based on the mmaction2, how can we obtain the current training epoch when computing the loss function? E.g.:
```python
@LOSSES.register_module()
class MyLoss(BaseWeightedLoss):
    def _forward(self, output, target, epoch):
        # compute the loss that needs epoch
```

I found that the epoch number is maintained by the module `mmcv/runner`. But how can we obtain the `runner.epoch`  in the `mmaction/models/losses` module?

**Comment by dreamerlin** [comment]
You may need to pass the `runner.epoch` [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/runner/epoch_based_runner.py#L50). @ycxioooong, any suggestions ?

**Comment by Cogito2012** [comment]
@dreamerlin That means the `mmcv` needs to be modified and re-installed. 
Could you explain more about what are the further modifications to pass the `runner.epoch` from the function `self.run_iter(data_batch, train_mode=True)` in mmcv to the `forward_train(self, imgs, labels, **kwargs)` of a recognizer in mmaction?

**Comment by Cogito2012** [comment]
@dreamerlin I came up with a solution and it will not modify the `mmcv`. This is done by following steps:
1. Add an annealing switcher in the config file: `annealing_runner = True`
2. Create a new Runner class `AnnealingRunner` that inherits the `EpochBasedRunner` from `mmcv` and register it in `mmaction/core/runner/__init__.py`. 
3. Modify the file `mmaction/apis/train.py`:
    - Import the new runner: `from ..core import AnnealingRunner`
    - Change the runner switch statement:
    ```python
    # Runner = OmniSourceRunner if cfg.omnisource else EpochBasedRunner
    if cfg.omnisource:
        Runner = OmniSourceRunner
    elif cfg.annealing_runner:
        Runner = AnnealingRunner  # add annealing runner support
    else:
        Runner = EpochBasedRunner
    ```
    - Add the annealing param into kwargs: 
    ```python
    if cfg.annealing_runner:
        runner_kwargs.update(annealing=True)
    runner.run(data_loaders, cfg.workflow, cfg.total_epochs, **runner_kwargs)
    ```
    
4. In the new `annealing_runner.py`:
    - Override the `train` function by passing `**kwargs` to `self.run_iter` function.
    ```python
       self.run_iter(data_batch, train_mode=True, **kwargs)  # in mmcv, here **kwargs is missing!!!
    ```
    - Override the `run_iter` function by adding the self.epoch into kwargs:
     ```python
      if 'annealing' in kwargs:
            kwargs.update(epoch=self.epoch)
      ```
5. In the file `mmaction/models/recognizers/base.py`, add the missing kwargs into `aux_info` in `train_step` function:
    ```python
       aux_info.update(kwargs)  # the kwargs is missing in train_step !!!
       losses = self(imgs, label, return_loss=True, **aux_info)
    ```
6. Finnaly, the epoch can be obtained from the customized loss:
    ```python
    @LOSSES.register_module()
    class MyLoss(BaseWeightedLoss):
        def _forward(self, output, target, **kwargs):
            # compute the loss that needs epoch
            assert 'epoch' in kwargs, "epoch number is missing!"
            epoch = kwargs['epoch']
    ```

In summary, the training epoch is passed into loss function by **first creating a new runner**, which inherits the `EpochBasedRunner` and **then updating the kwargs** along the functional calling process. This is a bit complicated, but hope it can help to improve mmaction2 for annealing training dependent on training epoch (or any other training hyper-parameters).
